### ISSUE TITLE
Some submissions with multi-word titles still have spaces in the saved filename

### ISSUE THREAD

**Issue by Eupeptic1** [Simon-Tesla/RaccoonyWebEx#119]
**Description**
When downloading a submission with a title that has multiple words in it, not all of the spaces between words are converted into underscores in the filename.  This happens on at least FurryNetwork and possibly on other sites as well.

**To Reproduce**
1. Visit a submission on FurryNetwork that has multiple words in the title - the more words, the better.  Example, sfw: https://furrynetwork.com/photo/93132/thats-not-how-chance-works/
2. Mouse-over the raccoon head and click the "Download" button. The image will download.
3. On the local filesystem, navigate to where Raccoony saved the file, and look at the filename.

Expected result: All spaces in the title have been converted to underscores in the filename, like:

`93132_That's_not_how_chance_works_by_awkore.jpg`

Actual result: There are still some spaces in the filename, especially towards the middle, like:

`93132_That's_not_how chance works_by_awkore.jpg`

**Environment**
Raccoony 1.2.5 + PR #113
Firefox 78.15.0esr, 64-bit, desktop
Debian Linux 10.11, amd64/x86_64, desktop

**Comment by Eupeptic1** [comment]
**Analysis**

`sanitizePath` in `app/download.ts` doesn't use a "global" modifier when replacing spaces with underscores, on line [117](https://github.com/Simon-Tesla/RaccoonyWebEx/blob/ee946d29f5c5d43205470db290e0f15b6f7a6ae2/app/download.ts#L117):

`pathPart = pathPart.replace(" ", "_");`

Changing that line to

`pathPart = pathPart.replace(/ /g, "_");`

seems to solve the problem.

I have only tested it on a couple of images at FurryNetwork.  Further testing on other sites to follow in the next few days.





  It does use the "global"

**Comment by Eupeptic1** [comment]
Looking at files I have downloaded in the past, this also shows up on e621, because the filename gets built from the first few tags, and some of the tags have spaces in them.  It doesn't seem to affect Weasyl or FA.

**Comment by Eupeptic1** [comment]
Since this change potentially affects all sites Raccoony supports, I ran it through all of the semi-automated testing, in Firefox, Chrome, and Edge.

**tl;dr:** I think it worked.  The change correctly affected 7 submissions on e621 and 1 on FurryNetwork, and correctly did not affect any of the other submissions in the semi-automated test list.

**Long version**

I added a few submissions at FurryNetwork to the previous list of submissions for semi-automated testing.  The revised list has been supplied via email.

I tested on Chrome, Edge, and Firefox.  For each browser, I first loaded stock 1.2.5 Raccoony, turned on auto-download and metadata files, and downloaded the test list.  I then moved the downloaded files away, cleared the browser download history, and unloaded stock 1.2.5.  I loaded Raccoony as modified to fix this bug, turned on auto-download and metadata files, downloaded the test list again, and compared the downloaded files in various ways.

**Versions used**
Google Chrome Version 96.0.4664.45 (Official Build) (64-bit)
Microsoft Edge Version 97.0.1072.8 (Official build) dev (64-bit)
Firefox 78.15.0esr (64-bit)

All of these are desktop, on Debian 10 desktop, x86_64.

**1. Chrome**

Smoke test: `find chrome-stock-1.2.5 -type f | grep ' '` lists files with spaces in their names, while `find chrome-with-fix -type f | grep ' '` lists no files.  Pass.

File count: 112 with stock 1.2.5 and 118 with the proposed fix.  When I ran the test with stock 1.2.5, I forgot that I had to log in to FA first, as some of the submissions on the test list are NSFW.  I *think* I forgot to log in to Weasyl for either test, because one of the submissions at Weasyl is NSFW; the counts *should* have been 114 and 120 respectively.

File name comparison (via `find . -type f | sort` and `diff`): WIth stock 1.2.5, 7 files at e621 and 1 file at FurryNetwork, plus their corresponding metadata files, have spaces in either the file name or directory name or both.  With the proposed fix, none of those files have spaces in the file name nor in the directory name.  Pass.

With the proposed fix, three files at FA, plus their corresponding metadata files, appear in the list, but do not appear in the list with stock 1.2.5.  I manually verified that these three files are rated other than General at FA, and are not available for download if you are not logged in.  I believe this explains the file count discrepancy.  Pass.

Content check: Out of the 230 files total, `fdupes -r -m chrome-stock-1.2.5 chrome-with-fix` reports `112 duplicate files (in 112 sets), occupying 102.9 megabytes`, so the submission and metadata files shared between the two sets are binary equal.  (The local disk filename does not appear in the metadata file, so this is expected.)  Pass.

I also generated the MD5s for both sets of files, did `cat chrome-stock-md5.txt chrome-fix-md5.txt | sort | uniq -c  | sort -n`, and reviewed the differences.  Of the 38 unique lines, 28 of them are for the 7 submission+metadata files at e621, and 4 of them are for 1 submission+metadata file at FurryNetwork.  The corresponding submission and metadata files have identical MD5s.  The remaining 6 unique lines are for the submission+metadata files that were downloaded in the "with proposed fix" pass, after I logged in to FA.  Pass.

**2. Edge**

Smoke test: `find edge-stock-1.2.5 -type f | grep ' '` lists files with spaces in their names, while `find edge-with-fix -type f | grep ' '` lists no files.  Pass.

File count: 119 with stock 1.2.5 and 120 with the proposed fix.  When I ran the test with stock 1.2.5, I had not shut off Edge's "open Word files in browser" feature, which interferes with Raccoony's file downloading.  I figured out how to work around it, but I lost one file while doing so.

File name comparison (via `find . -type f | sort` and `diff`): WIth stock 1.2.5, 7 files at e621 and 1 file at FurryNetwork, plus their corresponding metadata files, have spaces in either the file name or directory name or both.  With the proposed fix, none of those files have spaces in the file name nor in the directory name.  Pass.

With the proposed fix, one submission at FA appears on the list that does not appear on the list with stock 1.2.5.  This is the file that I lost while trying to work around Edge's "open Word files in browser" feature.  I believe this explains the file count discrepancy.  Pass.

Content check: Out of the 239 files total, `fdupes -r -m edge-stock-1.2.5 edge-with-fix` reports `119 duplicate files (in 119 sets), occupying 103.1 megabytes`, so the submission and metadata files shared between the two sets are binary equal.  (The local disk filename does not appear in the metadata file, so this is expected.)  Pass.

I also generated the MD5s for both sets of files, did `cat edge-stock-md5.txt edge-fix-md5.txt | sort | uniq -c  | sort -n`, and reviewed the differences.  Of the 33 unique lines, 28 of them are for the 7 submission+metadata files at e621, and 4 of them are for 1 submission+metadata file at FurryNetwork.  The corresponding submission and metadata files have identical MD5s.  The remaining 1 unique line is for the submission file that was downloaded in the "with proposed fix" pass, after I had worked around Edge's "open Word files in browser" feature.  Pass.

**3.  Firefox**

Smoke test: `find ff-stock-1.2.5 -type f | grep ' '` lists files with spaces in their names, while `find ff-with-fix -type f | grep ' '` lists no files.  Pass.

File count: 120 with stock 1.2.5 and 120 with the proposed fix.

File name comparison (via `find . -type f | sort` and `diff`): WIth stock 1.2.5, 7 files at e621 and 1 file at FurryNetwork, plus their corresponding metadata files, have spaces in either the file name or directory name or both.  With the proposed fix, none of those files have spaces in the file name nor in the directory name.  Pass.

Content check: Out of the 240 files total, `fdupes -r -m ff-stock-1.2.5 ff-with-fix` reports `120 duplicate files (in 120 sets), occupying 103.1 megabytes`, so the submission and metadata files shared between the two sets are binary equal.  (The local disk filename does not appear in the metadata file, so this is expected.)  Pass.

I also generated the MD5s for both sets of files, did `cat ff-stock-md5.txt ff-fix-md5.txt | sort | uniq -c  | sort -n`, and reviewed the differences.  Of the 32 unique lines, 28 of them are for the 7 submission+metadata files at e621, and 4 of them are for 1 submission+metadata file at FurryNetwork.  The corresponding submission and metadata files have identical MD5s.  Pass.

**4. Cross tab among browsers**

I compared the file lists from each browser when using stock 1.2.5: `cat chrome-stock.txt ff-stock.txt edge-stock.txt  | sort | uniq -c | sort -n`.  A "perfect score" would be 3 copies of each file.  There were 20 files that only had 2 or 1 copy.  Twelve of these were due to Chrome/Edge disallowing tildes in file and directory names; there were six unique files from Firefox with tildes, and six unique files from Chrome and Edge with the tildes replaced by underscores.  Six of these were due to me not logging in to FA during the first Chrome test, and two of these were due to me not logging in to Weasyl during either Chrome test.

I compared the file lists from each browser when using the proposed fix: `cat chrome-with-fix.txt ff-with-fix.txt edge-with-fix.txt  | sort | uniq -c | sort -n`.  A "perfect score" would be 3 copies of each file.  There were 14 files that only had 2 or 1 copy.  Twelve of these were due to Chrome/Edge disallowing tildes in file and directory names; there were six unique files from Firefox with tildes, and six unique files from Chrome and Edge with the tildes replaced by underscores.  Two of these were due to me not logging in to Weasyl during either Chrome test.

Comparing all 709 files downloaded by the three browsers with `fdupes -r -m` reports `589 duplicate files (in 120 sets), occupying 515.2 megabytes`.  A "perfect score" would have been 720 files downloaded, with 600 duplicate files in 120 sets.  I believe the "missing" files are due to the difficulties I noted above with not being logged in, and losing one file to Edge.

I compared the MD5s of all files downloaded by all six browser versions with a `sort | uniq -c | sort -n`.  A "perfect score" would have been 6 copies of each MD5+filename combination.  There were 52 MD5+filename combinations that had less than a perfect score.  12 combinations were due to file/directory tildes (Firefox) vs underscores (Chrome/Edge).  8 combinations were due to trouble with not downloading the same files on all test runs (being logged in, Edge).  

32 combinations were due to some file/directory names having spaces in them, and some not, which is the point of fixing this bug.  :D

**Conclusion**

I believe the proposed fix is working correctly.  Git goodness to follow.
